internal/controllers/synthesis/gc.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type podGarbageCollector struct {
+	client          client.Client
+	creationTimeout time.Duration
+}
+
+func NewPodGC(mgr ctrl.Manager, creationTimeout time.Duration) error {
+	c := &podGarbageCollector{
+		client:          mgr.GetClient(),
+		creationTimeout: creationTimeout,
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "podGarbageCollector")).
+		Complete(c)
+}
+
+func (p *podGarbageCollector) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	pod := &corev1.Pod{}
+	err := p.client.Get(ctx, req.NamespacedName, pod)
+	if err != nil || pod.DeletionTimestamp != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	logger = logger.WithValues("podName", pod.Name, "podNamespace", pod.Namespace)
+
+	// Pods can't progress after exiting 0
+	if pod.Status.Phase == corev1.PodSucceeded {
+		logger = logger.WithValues("reason", "Success")
+		return ctrl.Result{}, p.deletePod(ctx, pod, logger)
+	}
+
+	// Avoid waiting for the lease to expire for broken nodes
+	if delta := timeWaitingForKubelet(pod, time.Now()); delta > 0 {
+		if delta < p.creationTimeout {
+			return ctrl.Result{RequeueAfter: p.creationTimeout - delta}, nil
+		}
+		logger = logger.WithValues("reason", "ContainerCreationTimeout")
+		return ctrl.Result{}, p.deletePod(ctx, pod, logger)
+	}
+
+	// GC pods from deleted compositions
+	comp := &apiv1.Composition{}
+	comp.Name = pod.GetLabels()[manager.CompositionNameLabelKey]
+	comp.Namespace = pod.GetLabels()[manager.CompositionNamespaceLabelKey]
+	err = p.client.Get(ctx, client.ObjectKeyFromObject(comp), comp)
+	logger = logger.WithValues("compositionName", comp.Name, "compositionNamespace", comp.Namespace, "compositionGeneration", comp.Generation)
+	if errors.IsNotFound(err) || comp.DeletionTimestamp != nil {
+		logger = logger.WithValues("reason", "CompositionDeleted")
+		return ctrl.Result{}, p.deletePod(ctx, pod, logger)
+	}
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting composition resource: %w", err)
+	}
+
+	// GC pods from missing synthesizers
+	syn := &apiv1.Synthesizer{}
+	syn.Name = comp.Spec.Synthesizer.Name
+	err = p.client.Get(ctx, client.ObjectKeyFromObject(syn), syn)
+	logger = logger.WithValues("synthesizerName", syn.Name, "synthesizerGeneration", syn.Generation)
+	if errors.IsNotFound(err) {
+		logger = logger.WithValues("reason", "SynthesizerDeleted")
+		return ctrl.Result{}, p.deletePod(ctx, pod, logger)
+	}
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting synthesizer: %w", err)
+	}
+
+	// Pods from old syntheses
+	if !podIsCurrent(comp, pod) {
+		const gracePeriod = 250 * time.Millisecond
```

Comment:
Could informer drift be larger than this in underlays?

Reply:
Let me see if there's a better way to handle this, or at least add some comments

Reply:
Up'd it to 1 second just in case

Reply:
My take is that it's extremely unlikely informers will be behind by >1s and even if they are, recreating the pod isn't a terrible failure mode.